### PR TITLE
Leading and middle rest elements in tuple types

### LINKED ISSUE #39595: Support inferring rest args that are not at the end of the function's args
## Search Terms

variadic infer rest

## Suggestion

TypeScript 4.0 added support for "Variadic Tuple Types", the key relevant change being that spreads for tuple types can be done anywhere in a tuple, not just at the end. However it appears this does not extend to rest and `infer`.

It should be possible to infer an `...args` that is not at the end of a function (callable?) type.

## Use Cases

I use TypeScript API clients generated from Swagger/OpenAPI specs in React code by creating hooks to make the API calls in a React safe way. Naturally I want the to pass the API method args through the hook so the input args have the correct type.

However the API methods on these clients typically have something like an `options?: FetchOptionsType` at the end that shouldn't be part of the input args type. Options like these would be the purview of the hook itself, not relevant to the input arg. `(...args: infer A, options? FetchOptionsType) => any` seems like it would be the natural way to infer input args without the last arg, but does not work.

## Examples

```ts

type ApiMethod = (...args: any[]) => Promise<void>;
// This does not work, even in TS 4.0
type ApiMethodArgs<T> = T extends (...args: infer A, options?: any) => any
  ? A
  : never;
// This would work but gives the wrong type
// type ApiMethodArgs<T> = T extends (...args: infer A) => any
//   ? A
//   : never;
// @note The following works so infer is valid when mixing rest/non-rest args but just doesn't work if rest args are not last
// type ApiMethodArgs<T> = T extends (options: any, ...args: infer A) => any
//   ? A
//   : never;

// External API method, all API methods follow 
async function getUser(id: string, options?: any): Promise<void> { }

function callApi<AM extends ApiMethod>(method: AM) {
    const commonApiOptions = {};
    return async (...args: ApiMethodArgs<AM>): Promise<void> => {
        await method(...args, commonApiOptions)
    }
}

callApi(getUser)('asdf');
```

https://www.typescriptlang.org/play/index.html?ts=4.0.0-beta#code/FAFwngDgpgBAghAlgWSiAFgewCYwLwwAUAdKQIYBOA5gM4BcMZAdmANoC6AlPgHwwAKFTAFtENKAB4AbpkTYeAbmAB6ZTAAq6MTGyYoNGE0wgYAd0wUA1gBoYUKVCYxET9QGUYAFmIAGUJFgEFDQsbDhqGgl1PgJ1OwAPEEdsAxJyCIYXADMoCnhbTAgQREwmGgB+BmYwbjw+auAYGHL4RpgGJntcpVUNLQNzAFcAG1xzKxgAI0GTKkQHAwxYUyEmKhhwaBU1TcCkVAwccNoomI0EpKYUolJiSlpMphy8uFr6lm2m5tbepo6uig9NQAASMST6sCymGGw0wphc63GlgMNEwzieuWcBikZGGcjM6EcMFE8QRMAo+hAyiMTAAtBSaCZ7gZpiYAFaDRk6PQ0JgAchMSOcWXJlMYEXFsDBMGGZEZn128H2ISOEVO+HOUESyVShWKpXojBYtluzMez3gbyNYE+TRacFt7UMAKUnwAotqKExcfB+ABJYkq7C2XHDX0B4RBgxQmFwmDAOVgJgAYxgWUGKf1TioaAAquIKIQ5AxGRQEQUiiUypVrZwGIIRGJJDI5HwAN4wAC+wGA6czVZgydDQQkcGQFx1SuCh3khEjM4YY+4bbaTWTBpM6+EwlKQQA8pWDRq252lF9RSBBl7GDQk6m0ncMlODqFjpExzw6wIhKJxNJZPIvAwCu57nmQphkIgJjzqED7MrYW47kw+6HmUnCrl2wDdsAQ4wkEhA5iA+a5JwhB8nK2BZHynAKEAA

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

### PR DESCRIPTION
**Pull request by ahejlsberg**
With this PR we support leading and middle rest elements in tuple types. For example, we now support the following:

```ts
type T1 = [...string[], number];  // Zero or more strings followed by a number
type T2 = [number, ...boolean[], string, string];  // Number followed by zero or more booleans followed by two strings
```

Previously it was not possible to have anything follow a rest element and thus it was not possible to express tuple types that end in a fixed set of elements. Such tuple types are useful for strongly typing functions with variable parameter lists that end in a fixed set of parameters. For example, the following is now permitted:

```ts
function f1(...args: [...string[], number]) {
    const strs = args.slice(0, -1) as string[];
    const num = args[args.length - 1] as number;
    // ...
}

f1(5);
f1('abc', 5);
f1('abc', 'def', 5);
f1('abc', 'def', 5, 6);  // Error
```

Note the use of a rest parameter with a tuple type that _starts_ with a rest element. Also note that indexing a tuple type beyond its starting fixed elements (if any) yields a union type of all possible element types (in this case `string | number`). Thus, type assertions are required in the code above.

Tuple type layouts are now governed by these rules:

* An optional element cannot precede a required element or follow a rest element.
* Multiple rest elements are not permitted.

Thus, the supported layouts of tuple types are:

* Zero or more required elements, followed by zero or more optional elements, followed by zero or one rest element, or
* Zero or more required elements, followed by a rest element, followed by zero or more required elements.

In either layout, zero or more generic variadic elements may be present at any position in the tuple type. Errors are reported on tuple types that don't match the layout rules, but through generic type instantiation it is still possible to create invalid layouts. For this reason, tuple types are _normalized_ following generic type instantiation:

* Optional elements preceding the last required element are turned into required elements.
* Elements between the first rest element and the last rest or optional element are turned into a single rest element with a union of the element types.

Some examples of normalization:

```ts
type Tup3<T extends unknown[], U extends unknown[], V extends unknown[]> = [...T, ...U, ...V];

type TN1 = Tup3<[number], string[], [number]>;  // [number, ...string[], number]
type TN2 = Tup3<[number], [string?], [boolean]>;  // [number, string | undefined, boolean]
type TN3 = Tup3<[number], string[], [boolean?]>;  // [number, (string | boolean | undefined)[]]
type TN4 = Tup3<[number], string[], boolean[]>;  // [number, (string | boolean)[]]
type TN5 = Tup3<string[], number[], boolean[]>;  // (string | number | boolean)[]
```

Fixes #39595.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this

**Comment by ahejlsberg** [comment]
All tests look clean and perf appears unaffected.

**Comment by IllusionMH** [comment]
```ts
type TN2 = Tup3<[number], [string?], [boolean]>;  // [number, string | undefined, boolean]
```
but `[]` is also assignable to `[string?]` shouldn't it be like `[number, string | undefined, boolean] | [number, boolean]`?
Or am I missing something?

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this

**Review by weswigham — COMMENTED** [review]
I'm going to pause for a second here to post what I have thus far, since it's mostly design feedback - specifically around the premise of this PR. The first line of the description is:

> With this PR we support leading and middle rest elements in tuple types

And that's a really good thing. Except as implemented right here, we only _actually_ support them in generic cases. And then we always end up erasing middle/leading spreads once the generics are gone into a much broader type. If we were erasing them into a much more specific type (so the result allowed fewer assignments), I could maybe get it - but the significantly reduced accuracy of these constructs once we're talking about non-generic types really rubs me the wrong way. I think we need to preserve things like `[...string[], ...number[]]` because those types have significant meaning - the same way the non-generic pattern literals we just added have meaning in the sequences they preserve.

File: tests/baselines/reference/spliceTuples.errors.txt [hunk 1]
```diff
@@ -28,5 +28,5 @@ tests/cases/compiler/spliceTuples.ts(23,1): error TS2322: Type '[number, string,
     k6 = [1, ...sbb_];
     ~~
 !!! error TS2322: Type '[number, string, boolean, ...boolean[]]' is not assignable to type '[number, string, boolean, boolean, ...boolean[]]'.
-!!! error TS2322:   Property '3' is optional in type '[number, string, boolean, ...boolean[]]' but required in type '[number, string, boolean, boolean, ...boolean[]]'.
+!!! error TS2322:   Source provides no match for required element at position 3 in target.
```
[the remark this answers is not in the document]
Comment (DanielRosenwasser) [reply tests/baselines/reference/spliceTuples.errors.txt:31]: https://github.com/microsoft/TypeScript/issues/41860 now keeps track of some ideas we have in mind here.

**Comment by DanielRosenwasser** [comment]
@weswigham from your comments, it sounds like you want to loosen the restriction in place:

> Tuple type layouts are now governed by these rules:
> 
> * An optional element cannot precede a required element or follow a rest element.
> * Multiple rest elements are not permitted.
_________

For optional elements, I tend to somewhat agree. As a user, I would expect that we might desugar

```ts
type Foo = [string, string?, number];
```

into something like

```ts
type Foo = [string, number] | [string, string, number];
```

though not the suggested

```ts
type Foo = [string, number] | [string, string | undefined, number];
```

How often do we leverage the distinction between optional and potentially undefined when we have the property itself?

_____________

For multiple rest elements... well, I don't have a strong-enough intuition on that one, but it would be pretty different in behavior. It would line up better with what we have in template string types.

**Comment by weswigham** [comment]
> it sounds like you want to loosen the restriction in place:

I guess so. Since you can replicate those situations in cases where we do not issue an error (eg, via variadic concatenation), I think having accurate behavior for them is important. And thus, once you have accurate behavior for those situations, there's not much of a justification in forbidding them in normal circumstances, I think.


> How often do we leverage the distinction between optional and potentially undefined when we have the property itself?

Infrequently. In fact, that we don't is pretty much one of the reasons for https://github.com/microsoft/TypeScript/issues/41418 being a bug and not "intended behavior". So the inclusion of `undefined` in the type is potentially needed for consistency.

> For multiple rest elements...

Yeah, I just reason that sequences of types in an array aren't that different from sequences of characters (or types) in a string (or template) - so the later being strictly more powerful with respect to what is preserved through concatenation operations rubs me the wrong way a bit - I feel like we've done some of the thinking on how to match up these sequences, and should probably try to do the same here, rather than quietly losing type fidelity in an unsafe way.

**Review by janyawat — COMMENTED** [review]
leave a comment

**Review by janyawat — COMMENTED** [review]
leave a comment

**Pull request merged**